### DISCUSSION TITLE
HARP motor control / HID output

### DISCUSSION THREAD

**Discussion by ederancz** [neurogears/vestibular-vr#16] [Q&A]
Hello, 

I have a question regarding the output from the HARP board. How will the optical sensor data be communicated? Is it possible to get a HID-type output? 

I am asking because we are building a simpler visual stimulation platform based on this (https://github.com/sn-lab/mouseVRheadset) and psychopy. 

The idea is to 1. use a simpler and cheaper visual stimulus delivery for simple experiments before moving the mouse onto the multi-screen VR on a second rotation platform, and 2. start experiments in a month or so. 

cheers,
Ede

**Comment by RoboDoig** [comment]
Hi Ede,

At the moment the optical sensor data is read from the device as an array of int16 values [xDelta1, yDelta1, brightness1, xDelta2, yDelta2, brightness2] which are both parsed in Bonsai and transmitted directly to the motor control board via the slip-ring. 

I'm not that familiar with the requirements for a HID (this refers to human-interface device?) output but from a quick read of its description it seems like we could probably package the data packet into the appropriate format.

Is the idea that the HARP device would be read as a HID device over USB, communicating the optical flow x, y?

**Comment by ederancz** [comment]
Thanks Andrew, 

The idea is to have a parallel output of the optical sensors go to a separate visual stimulus delivery device on the platform. The reason for this parallel development is that [this](https://github.com/sn-lab/mouseVRheadset) will be ready to go in a few weeks, and we are keen to start experiments (we will do this on the original platform, but it would be nice to be compatible with the HARP). 

One simple solution could be a separate optical sensor, including its control board, and feed that USB to the Raspberry Pi of the visual stimulus device. The practical problem with that could be consistency between optical sensors, such as the visual and motor commands are the same.

**Comment by RoboDoig** [comment]
OK - I understand a bit better now I think. Since we have direct tx/rx output on the sensor board, we should be able to have a parallel stream to the Pi. I will have a think about other potential solutions this week.

**Reply by ederancz** [reply]
Great, thanks! The Pi will be on the platform though, so an output from the onboard HARP would be best, alternatively through the slipring of course.

**Reply by RoboDoig** [reply]
We had a discussion about this today and there are a few potential solutions:

1. Pass the data communication (tx/rx) port of the HARP H1 to an Arduino Due which can be setup as a HID device that connects to the Pi. Would require a little extra development on the Arduino side and an extra board on the platform.
2. Have parallel HARP board firmware that can be loaded onto the board (depending which visual stimulus rig is used) that sets up the HARP device as a HID. HARP connects directly to Pi. This would require more firmware development and would probably be the most time intensive.
3. As you mentioned, use a separate optical sensor (e.g. gaming mouse) as input to the Pi. Would require some extra comparison to the optical flow sensors to assess consistency between different sensor types.
4. We recently integrated ZeroMQ networking into Bonsai. We could read the HARP device motion data as we do now in Bonsai and transmit the motion signal back to the Pi over network, e.g. with a client running in Godot that receives motion signal.  Advantage here is very easy setup, but would need to characterise network delays / jitter.
5. Have Bonsai running on the Pi (or within Godot) directly reading the HARP device.

I think options 1) or 4) would be the easiest to get up and running quickly. Would you want the two visual stimulus devices to be present on the same platform? And would the VR goggle also be used with the platform rotating?

Let me know if it would be useful to schedule a meeting to discuss.
